synapse/crypto/keyring.py

```diff
@@ -420,26 +449,22 @@ async def _inner_fetch_key_request(
                 if not key:
                     continue
 
-                # If we already have a result for the given key ID we keep the
+                # If we already have a result for the given key ID, we keep the
                 # one with the highest `valid_until_ts`.
                 existing_key = found_keys.get(key_id)
-                if existing_key:
-                    if key.valid_until_ts <= existing_key.valid_until_ts:
-                        continue
+                if existing_key and existing_key.valid_until_ts >= key.valid_until_ts:
+                    continue
```

Comment:
By going from `a <= b` to `b >= a` I think we've swapped what `a == b` does.

Reply:
Hrm, indeed. That serves me right for trying to eyeball boolean logic on the fly. Have fixed in [357eef8](https://github.com/matrix-org/synapse/pull/14804/commits/357eef8982cedf2047db338112b60c8be3f94101).